lib/auth/auth_login_test.go

```diff
@@ -701,6 +702,111 @@ func TestServer_Authenticate_nonPasswordlessRequiresUsername(t *testing.T) {
 	}
 }
 
+func TestServer_Authenticate_headless(t *testing.T) {
+	t.Parallel()
+	srv := newTestTLSServer(t)
+
+	// We don't mind about the specifics of the configuration, as long as we have
+	// a user and TOTP/WebAuthn devices.
+	mfa := configureForMFA(t, srv)
+	username := mfa.User
+
+	proxyClient, err := srv.NewClient(TestBuiltin(types.RoleProxy))
+	require.NoError(t, err)
+
+	headlessID := services.NewHeadlessAuthenticationID([]byte(sshPubKey))
+	ctx := context.Background()
+
+	timeout := time.Millisecond * 200
+
+	updateHeadlessAuthnInGoroutine := func(ctx context.Context, update func(*types.HeadlessAuthentication)) chan error {
+		errC := make(chan error)
+		go func() {
+			defer close(errC)
+
+			headlessAuthn, err := srv.Auth().GetHeadlessAuthentication(ctx, headlessID)
+			if err != nil {
+				errC <- err
+				return
+			}
+
+			// create a shallow copy and update for the compare and swap below.
+			replaceHeadlessAuthn := *headlessAuthn
+			update(&replaceHeadlessAuthn)
+
+			_, err = srv.Auth().CompareAndSwapHeadlessAuthentication(ctx, headlessAuthn, &replaceHeadlessAuthn)
+			if err != nil {
+				errC <- err
+				return
+			}
+		}()
+		return errC
+	}
+
+	for _, tc := range []struct {
+		name     string
+		update   func(*types.HeadlessAuthentication)
+		checkErr require.ErrorAssertionFunc
+	}{
+		{
+			name: "OK approved",
+			update: func(ha *types.HeadlessAuthentication) {
+				ha.State = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED
+				ha.MfaDevice = mfa.WebDev.MFA
+			},
+			checkErr: require.NoError,
+		}, {
+			name: "NOK approved without MFA",
+			update: func(ha *types.HeadlessAuthentication) {
+				ha.State = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED
+			},
+			checkErr: require.Error,
+		}, {
+			name: "NOK user mismatch",
+			update: func(ha *types.HeadlessAuthentication) {
+				ha.State = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED
+				ha.MfaDevice = mfa.WebDev.MFA
+				ha.User = "other-user"
+			},
+			checkErr: require.Error,
+		}, {
+			name: "NOK denied",
+			update: func(ha *types.HeadlessAuthentication) {
+				ha.State = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_DENIED
+			},
+			checkErr: require.Error,
+		}, {
+			name:     "NOK timeout",
+			update:   func(ha *types.HeadlessAuthentication) {},
+			checkErr: require.Error,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Cleanup(func() {
+				srv.Auth().DeleteHeadlessAuthentication(ctx, headlessID)
+			})
```

Comment:
Any reason why we delete data from the server instead of creating a new one for each test? Currently, the state of the server is shared between test cases. Potentially we could also run each test case in a separate goroutine to make it faster.